rasa/core/policies/embedding_policy.py

```diff
@@ -1102,240 +959,298 @@ def _linearly_increasing_batch_size(self, epoch: int) -> int:
+                    )
+                    batches_per_epoch += 1
+                    ep_train_loss += batch_train_loss
+                    ep_train_acc += batch_train_acc
 
-    # noinspection PyPep8Naming
-    def _scale_loss_by_count_actions(
-        self,
-        X: np.ndarray,
-        slots: np.ndarray,
-        previous_actions: np.ndarray,
-        actions_for_Y: np.ndarray,
-    ) -> Union[np.ndarray, List[List]]:
-        """Calculate inverse proportionality of repeated actions."""
-
-        if self.scale_loss_by_action_counts:
-            full_X = np.concatenate(
-                [X, slots, previous_actions, actions_for_Y[:, :, np.newaxis]], -1
+                except tf.errors.OutOfRangeError:
+                    break
+
+            train_loss = ep_train_loss / batches_per_epoch
+            train_acc = ep_train_acc / batches_per_epoch
+
+            pbar.set_postfix(
+                {"loss": "{:.3f}".format(train_loss), "acc": "{:.3f}".format(train_acc)}
             )
-            full_X = full_X.reshape((-1, full_X.shape[-1]))
 
-            _, i, c = np.unique(full_X, return_inverse=True, return_counts=True, axis=0)
+            if eval_init_op is not None:
+                if (ep + 1) % self.evaluate_every_num_epochs == 0 or (
+                    ep + 1
+                ) == self.epochs:
+                    eval_loss, eval_acc = self._output_training_stat_dataset(
+                        eval_init_op, loss, acc
+                    )
+                    if (ep + 1) != self.epochs:
+                        logger.info(
+                            "Evaluation results: "
+                            "validation loss: {:.3f}, "
+                            "validation accuracy: {:.3f}"
+                            "".format(eval_loss, eval_acc)
+                        )
+
+        final_message = (
+            "Finished training embedding policy, "
+            "train loss={:.3f}, train accuracy={:.3f}"
+            "".format(train_loss, train_acc)
+        )
+        if eval_init_op is not None:
+            final_message += (
+                ", validation loss={:.3f}, validation accuracy={:.3f}"
+                "".format(eval_loss, eval_acc)
+            )
+        logger.info(final_message)
+
+    def _output_training_stat_dataset(
+        self, eval_init_op: "tf.Operation", loss: "tf.Tensor", acc: "tf.Tensor"
+    ) -> Tuple[float, float]:
+        """Output training statistics"""
+
+        self.session.run(eval_init_op)
+        ep_val_loss = 0
+        ep_val_acc = 0
+        batches_per_epoch = 0
+        while True:
+            try:
+                batch_val_loss, batch_val_acc = self.session.run(
+                    [loss, acc], feed_dict={self._is_training: False}
+                )
+                batches_per_epoch += 1
+                ep_val_loss += batch_val_loss
+                ep_val_acc += batch_val_acc
+            except tf.errors.OutOfRangeError:
```

Comment:
That seems kind of hacky. Can you find a better condition for interrupting the while loop?

Reply:
this is how they recommend to do it in tf docs: https://www.tensorflow.org/guide/datasets#consuming_values_from_an_iterator